sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,745 @@
+
+                centroid_window_first_loading_feature_idx = 0
+
+                # The work items in the work group load cooperatively the values in
+                # shared memory. At each iteration, the work item loads one value and
+                # adjacent work items load adjacent values.
+                for _2 in range(centroids_window_height_ratio_multiplier):
+                    window_loading_feature_idx = (
+                        centroid_window_first_loading_feature_idx
+                        + window_loading_feature_offset
+                    )
+                    loading_feature_idx = first_feature_idx + window_loading_feature_idx
+
+                    if (loading_feature_idx < n_features) and (
+                        loading_centroid_idx < n_clusters
+                    ):
+                        value = current_centroids_t[
+                            loading_feature_idx, loading_centroid_idx
+                        ]
+                    else:
+                        value = zero
+
+                    centroids_window[
+                        window_loading_feature_idx, window_loading_centroid_idx
+                    ] = value
+
+                    centroid_window_first_loading_feature_idx += (
+                        nb_of_window_features_per_work_group
+                    )
+
+                # Since other work items are responsible for loading the relevant data
+                # for the next step, we need to wait for completion of all work items
+                # before going forward
+                dpex.barrier()
+
+                # Loop on all features in the current window and accumulate the dot
+                # products
+                for window_feature_idx in range(window_nb_features):
+                    feature_idx = window_feature_idx + first_feature_idx
+                    if (feature_idx < n_features) and (sample_idx < n_samples):
+                        # performance for the line thereafter relies on L1 cache
+                        X_value = X_t[feature_idx, sample_idx]
+                    else:
+                        X_value = zero
+                    # For this given feature, loop on all centroids in the current
+                    # window and accumulate the dot products
+                    for window_centroid_idx in range(window_nb_centroids):
+                        centroid_value = centroids_window[
+                            window_feature_idx, window_centroid_idx
+                        ]
+                        partial_dot_products[window_centroid_idx] += (
+                            centroid_value * X_value
+                        )
+
+                # When the next iteration starts work items will overwrite shared memory
+                # with new values, so before that we must wait for all reading
+                # operations in the current iteration to be over for all work items.
+                dpex.barrier()
```

Comment:
We only need a local fence here, right?

Reply:
Woops, big mistake here, I thought it was local by default, [apparently it's not](https://intelpython.github.io/numba-dpex/latest/user_guides/kernel_programming_guide/synchronization.html).

All the barriers in the current code are meant to be local.

It might be a good news for performance :-D

Reply:
The performance since to decrease after this change :thinking: